src/mscorlib/shared/System/String.cs

```diff
@@ -2,109 +2,391 @@
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using System.Text;
 
 namespace System
 {
-    using System.Text;
-    using System;
-    using System.Buffers;
-    using System.Runtime;
-    using System.Runtime.ConstrainedExecution;
-    using System.Globalization;
-    using System.Threading;
-    using System.Collections;
-    using System.Collections.Generic;
-    using System.Runtime.CompilerServices;
-    using System.Runtime.InteropServices;
-    using System.Runtime.Versioning;
-    using Microsoft.Win32;
-    using System.Diagnostics;
-    using System.Security;
-
-    //
-    // For Information on these methods, please see COMString.cpp
-    //
     // The String class represents a static string of characters.  Many of
     // the String methods perform some type of transformation on the current
-    // instance and return the result as a new String. All comparison methods are
-    // implemented as a part of String.  As with arrays, character positions
-    // (indices) are zero-based.
+    // instance and return the result as a new String.  As with arrays, character
+    // positions (indices) are zero-based.
 
     [Serializable]
     [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
-    public sealed partial class String : IComparable, ICloneable, IConvertible, IEnumerable
-        , IComparable<String>, IEnumerable<char>, IEquatable<String>
+    public sealed partial class String : IComparable, IEnumerable, IEnumerable<char>, IComparable<String>, IEquatable<String>, IConvertible, ICloneable
     {
-        //
-        //NOTE NOTE NOTE NOTE
-        //These fields map directly onto the fields in an EE StringObject.  See object.h for the layout.
-        //
-        [NonSerialized] private int m_stringLength;
+        // String constructors
+        // These are special. The implementation methods for these have a different signature from the
+        // declared constructors.
 
-        // For empty strings, this will be '\0' since
-        // strings are both null-terminated and length prefixed
-        [NonSerialized] private char _firstChar;
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        public extern String(char[] value);
 
-        // The Empty constant holds the empty string value. It is initialized by the EE during startup.
-        // It is treated as intrinsic by the JIT as so the static constructor would never run.
-        // Leaving it uninitialized would confuse debuggers.
-        //
-        //We need to call the String constructor so that the compiler doesn't mark this as a literal.
-        //Marking this as a literal would mean that it doesn't show up as a field which we can access 
-        //from native.
-        public static readonly String Empty;
+#if PROJECTN
+        [DependencyReductionRoot]
```

Comment:
This is needed because it is called by ProjectN runtime?

Reply:
Yes, this attribute is needed on this by the ProjectN toolchain.